### PR TITLE
Create poll module

### PR DESCRIPTION
**Pull request by AntoineJT**
# Guide
### Configure the module
image
You don't have to configure the module for it to work : people with the administrator permission can already create polls, specify channel, etc.

**It has 3 types of parameters:**
- **Permission**: One or more roles allowed or disallowed to do something. 
Please note that **members with the administrator permission are bypassing these rules**.
- **Default**: This can be overriden when initializing a poll.
- **Behavior**: This cannot be overriden and is affecting all the polls.
---
#### Parameters
Most of the informations are already on the screenshot. This is just like a minimal wiki-like thing.

- #### AllowedRoles
**Description:** Roles allowed to create polls
**Type:** Permission

- #### SpecifyChannelAllowedRoles
**Description:** Roles allowed to specify where to send a poll
**Type:** Permission

- #### DefaultPollChannel
**Description:** Where should polls be sent if no channel is set on init
**Type:** Default

- #### DefaultPollDuration
**Description:** Default poll duration if no duration is set on init
**Type:** Default

- #### DeletePollOnExpiration
**Description:** Delete original poll message on expiration
**Type:** Behavior

### How to create a poll
I will not explain everything you can do with this poll module, because no one would read that I think.

**Step 0:** Enable the module by using the command `enable poll`
image

**Step 1:** Initiate a poll by using the command `initpoll "<title>" [channel] [duration]`
**Don't forget to put your title between double quotes `""` or else only the first word will be set as a title!**
image

**Step 2:** Add at least 2 choices to the poll using the command `poll add <emoji> <text>`
You can remove choices by using the command `poll remove <emoji>`
image
You can note that here, no double quote is needed!

**Step 3:** When you're done with setting up your poll, you can send it by using the command `poll send`
image
Here you can see an example of poll!

**Step 4:** Wait for results!
By default, poll is deleted when results are sent!
image
Please note that **choices with the same number of votes are put in a random order!**
Here we can see that someone thinks the product will fail to be on the trend and someone else doesn't know! Nobody thinks it will be on the trend!

Now that you know how to basically use the module, you can read the informations about the different commands below.

### Commands
Here is some basic help about the commands.
- #### initpoll
image
Usage: `initpoll "<title>" [channel] [duration]`
**Don't forget to put your title between double quotes "" or else only the first word will be set as a title!**
- #### deinitpoll
image
Usage: `deinitpoll`
- #### poll
image
image
**All the different possible values of `action` are listed below:**
- **add**: Adds a choice
Usage: `poll add <emoji> <text>`
- **remove**: Removes a choice
Usage: `poll remove <emoji>`
- **update**: Updates text associated with an emoji
Usage: `poll update <emoji> <text>`
- **title**: Changes the title
Usage: `poll title <text>`
Here, no double quotes are needed!
- **send**: Sends the poll
Usage: `poll send`
---
# Complete history
* Implements startpoll command

* Some more work

* More work again

* More work

* Fix default poll channel, add todos

* Some improvements

- Polls storage has changed a bit allowing it (at least in theory) to avoid emojis to be sorted by their code : it's keeping order now
- Avoid some 400 Bad Request : ShowPoll is now FormatPoll, replying 1 message instead of 2 too quickly

* Some more work

* Remove useless title nil check

* Fix send command being broken

* Fix SendNotImplemented, some more work

* Replace ugly text by an embed

* Add reactions to preview embed, write preview in the title

* Some work again

* Handle the case when an emoji is deleted during the configuration of a poll

* Make it possible to send a poll

* Store only id of guild and members instead of the whole objects

* Fix stupid bugs

* Remove "Ignore the above warning" because the warning doesn't exists anymore

* Begin to work on the duration feature, remove useless guild id from persistant storage

The persistant storage is already made by guild, so it has no need to store it in

* Limit choices to 20 which is the reactions limit

* Cleanup a bit

* Cleanup a bit code and TODOs

* Add poll module to AutoloadModules in default config file

* Some work

* Send results when a poll is finished

* Fix results footer being ugly, remove duration warn

* Try to put choices with the same number of votes at the same position

This is broken for an unknown reason

* Revert "Try to put choices with the same number of votes at the same position"

This reverts commit 130f4427b2838a2f94e8b1e766981cc68151177d.

* Implements update action, some improvements

* Major improvements

* Remove some useless variables

* Try some things

* Improve results style

* Some broken shit

* Some broken shit again

* Use merge sort but it's still broken

The merge sort seems to be good, but the order is still bad : maybe the fields and the reactions are not in the same order when merged in a map

* Make it works!!

* Replace merge sort by lua native sort

* Remove useless do end block

* Some changes

### CODE REVIEW

File: module_poll.lua [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+		end)
+    end)
+
+    function IsAllowedToSpecifyChannel(channel, member, config)
+        return member:hasPermission(enums.permission.administrator) 
+            or (config.SpecifyChannelAllowedRoles ~= nil 
+                and HasOneOfTheseRoles(member, config.SpecifyChannelAllowedRoles))
+    end
+
+    function FormatVotes(count)
```
Comment (SirLynix) [inline module_poll.lua:178]: Use Module functions (like `Module:FormatVotes`) or local functions.

File: module_poll.lua [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+end
+
+function Module:OnUnload()
+	if (self.Clock) then
+		self.Clock:stop()
+	end
+end
+
+function Module:OnLoaded()
+    polls = {}
```
Comment (SirLynix) [inline module_poll.lua:87]: Don't use a global table for storing polls, use module data (or persistent data if required)

File: module_poll.lua [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+
+            if action == "add" then
+                if #polls[guild.id][member.id].choices >= 20 then
+                    commandMessage:reply("You can't add more than 20 choices!")
+                    return
+                end
+
+                local reply = nil
+                
+                if text == nil then
```
Comment (SirLynix) [inline module_poll.lua:371]: + add test if text exists but is empty

File: module_poll.lua [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+            function IsAChoice(emoji)
+                for _, choice in ipairs(polls[guild.id][member.id].choices) do
+                    if choice.emoji.Id == emoji.Id then
+                        return true
+                    end
+                end
+                return false
+            end
+
+            if polls[guild.id][member.id] == nil then
```
Comment (SirLynix) [inline module_poll.lua:358]: Use locals to prevent fetching for a poll everytime
